### DISCUSSION TITLE
Text in .txt in horizontally

### DISCUSSION THREAD

**Discussion by Jorg3cSs** [ImageMagick/ImageMagick#3979] [Help] [answered]
Hi.
I want to converter all pixels of a image to #hex 6 digits code in a Output.txt

I have this code:
`magick Test4x3.png rgb:- | xxd -g3 -c3 | awk '{print $2}' >> Output.txt`

And this is the Output.txt:
> ff0000
ff0000
ff0000
ff0000
0000ff
0000ff
0000ff
0000ff
00ff00
00ff00
00ff00
00ff00

The problem is with the Output.txt, all #hex codes is writen in vertically, and **i want all the text in horizontally.** 
This is what i want in the Output.txt:
> ff0000ff0000ff0000ff00000000ff0000--->

**Comment by snibgo** [comment]
If your input is only 4x3 pixels, I suggest:
```
magick Test4x3.png -crop 1x1 -depth 8 -format %[hex:p{0,0}] info:
```

**Reply by Jorg3cSs** [reply]
@snibgo 
`$ magick Test4x3.png -crop 1x1 -depth 8 -format %[hex:p{0,0}] info:`

And i get this:
> magick: unable to open image '%[hex:p0]': No such file or directory @ error/blob.c/OpenBlob/3537.
magick: no decode delegate for this image format `' @ error/constitute.c/ReadImage/572.

And my input is only 4x3 in this example, i want to do that with more resolution images.

**Reply by fmw42** [reply]
try putting double quotes about %[hex:...]

`magick Test4x3.png -crop 1x1 -depth 8 -format "%[hex:p{0,0}]" info:`

What version of ImageMagick and on what platform/OS?

**Reply by Jorg3cSs** [reply]
It works! 
But i want the output in a .txt file, is possible?

I don't know what version i have, how can i know that?

**Comment by snibgo** [comment]
"magick -version" tells you the version.

If you want to write to a file, put the filename immediately after "info:".
```
magick Test4x3.png -crop 1x1 -depth 8 -format "%[hex:p{0,0}]" info:myfile.txt
```

**Reply by fmw42** [reply]
`magick -version`

**Reply by Jorg3cSs** [reply]
`$ magick -version`
> Version: ImageMagick 7.1.0-4

Platform: Android v 10
EMUI 10.0.0

Works like a champ! 
Thank you very much for your help! 👍

**Reply by Jorg3cSs** [reply]
@snibgo @fmw42
Hi again. 
Today i tried this with a image 1600x1000, i put the code and when i press enter nothing happend, the $ symbol at the beginning dissapear...
And i have to restart termux because any command i write don't work without the $ at the beginning.

**Comment by snibgo** [comment]
"-crop 1x1" chops the image into many images, one image per input pixel. This is fine if the input has 12 pixels or a few thousand, but a million images will probably exceed your memory.

Your original command works fine, but you want to strip the newlines. Is that correct? Then use a Unix tool that does that job. For example, "tr -d \n".

**Reply by Jorg3cSs** [reply]
Yes, i think it is for memory exceed. 

Correct, i want all the pixel inputs in one text line horizontally.

Like with the original code and a 4x3 image, but i want to do this with a 1600x1000 image. 

I don't have a pc right now for try that Unix tool.

**Comment by Jorg3cSs** [comment]
@snibgo, 
I've search info about manipulating a large image like 2000x2000px in imagemagick with exceed memory problem, and i see  that for do this there's to edit the policy.xml file, - width, 
-height settings.

I would have to edit that file in order to get the hex code for a 2000px as max image?

I would like to use the code:

`magick Image1800px.png -crop 1x1 -depth 8 -format "%[hex:p{0,0}]"`

for images from a 2000px max to down. 

Is possible this editing policy.xml?

**Comment by snibgo** [comment]
Each image that IM stores in memory needs space for the pixel cache, plus an overhead of about 26000 bytes. The "-crop 1x1" operation writes each input pixel into its own image. So if you have 1 million pixels, they become one million images. The overhead for one million images is about 26 billion bytes, 26 GB.

If you don't have enough physical memory, the pixel caches will be stored on disk, but not the overhead. If you don't have enough memory, no changes to policy.xml will provide the memory.

So, as I say above, I do not recommend "-crop 1x1" with large images.

Instead, use your original command with a shell filter that strips newlines, such as "tr -d \n".

**Comment by Jorg3cSs** [comment]
@snibgo
mmm, ok, i understand.
Is a bunch of images that stores in memory...
$ magick -list resource

> Resource limits:
>   Width: 461.169PP
>   Height: 461.169PP
>   Area: 11.8844GP
>   List length: unlimited
>   Memory: 5.53412GiB
>   Map: 11.0682GiB
>   Disk: unlimited
>   File: 24576
>   Thread: 1
>   Throttle: 0
>   Time: unlimited

I will not edit the policy.xml.

Can you say me the original code without "-crop 1x1" but with the shell filter that you says me "tr -d \n", please? 

Im newbie with termux and imageshack. 

I try adding alone "tr -d \n", but i don't put exactly inthe place well the letters and only see errors.

I type "tr --help" and a lot of commands and info appears, but i don't understand it...

**Comment by snibgo** [comment]
In your OP, you showed this command:
```
magick Test4x3.png rgb:- | xxd -g3 -c3 | awk '{print $2}' >> Output.txt
```
The output is almost what you want, but you want to remove the newlines.

I suggest you try this command:
```
magick Test4x3.png rgb:- | xxd -g3 -c3 | awk '{print $2}' | tr -d \n >> Output.txt
```

**Comment by Jorg3cSs** [comment]
Thank you.
But the output text of this code is vertically
This is the code for horizontal text:

`magick Test4x3.png -crop 1x1 -depth 8 -format "%[hex:p{0,0}]" info:Output.txt`

Can you  add "tr -d \n" and delete "-crop 1x1" to  that code please?
Thanks in andvance.

**Comment by snibgo** [comment]
> Can you add "tr -d \n" and delete "-crop 1x1" to that code please?

That won't work, because "-format" is run only once per image. That is why "-format" needs "-crop 1x1", to crop each pixel into its own image.

**Comment by fmw42** [comment]
In unix, you can do the following using txt: format so that you only read the image once and no crops.

`convert rose:[10x10+30+40] txt: | sed 's/^.*#\([^ ]*\).*$/\1/g' | tr -d "\n"`

C74839A84536874237803F36823B327D393278382F75372E73372E6F352BBD392EB63C39983A3B7F403D70423C714036713F356F3F346A3C32683A31B46B539D3F2E91362A7F382C6F4133723D37733F3973413B6F3F386E3D36FFFEEBDAC3ACA0755D7948316536226533296A39306C3F34693E34683E32FFFFFFFFFFFFFFFFFEF8F4E4E5D8C8D1BDA768533B58442C5D4B325C4C34FEFAFBFEFFFDFDFFFFFFFFFCFFFFEEE9E6CD6D674B524B2D554F32534F33

which is actually one long line.  Replace rose:[10x10+30+40] with just image.suffix for your image.

**Comment by Jorg3cSs** [comment]
Thanks for your reply @fmw42 

Well! This code:

`convert Test4x3.png txt: | sed 's/^.*#\([^ ]*\).*$/\1/g' | tr -d "\n"`

works with 1200px images! Nice! 
And i suppose this too work with more than 1200px images!
And no memory exceed problem! 👍

It gives me all the code in horizontally:

`FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF80C01F80C01F80C01F80C01`

(that is what i want), but in the terminal.

If it is a low resolution image i can copy all the text and paste in a .txt file. 

But if is a 1200px image, how can i copy all the text? a super long text log, i want to scroll up back to the beginning but the scrolling ends just halfway..or scrolling up while copy... 

Or Output to a .txt file? 
I try this:

`convert Test4x3.png Output.txt: | sed 's/^.*#\([^ ]*\).*$/\1/g' | tr -d "\n"`

But this is the Output.txt:

>  ImageMagick pixel enumeration: 4,3,255,srgb
> 0,0: (255,255,255)  #FFFFFF  white
> 1,0: (255,255,255)  #FFFFFF  white
> 2,0: (255,255,255)  #FFFFFF  white
> 3,0: (255,255,255)  #FFFFFF  white
> 0,1: (255,255,255)  #FFFFFF  white
> 1,1: (255,255,255)  #FFFFFF  white
> 2,1: (255,255,255)  #FFFFFF  white
> 3,1: (255,255,255)  #FFFFFF  white
> 0,2: (248,12,1)  #F80C01  srgb(248,12,1)
> 1,2: (248,12,1)  #F80C01  srgb(248,12,1)
> 2,2: (248,12,1)  #F80C01  srgb(248,12,1)
> 3,2: (248,12,1)  #F80C01  srgb(248,12,1)

I don't want that format, i want only the 6 hex digits (ff00ff) in horizontally.

**Comment by snibgo** [comment, marked as answer]
You can redirect the text to a new file by adding this to the end:
```
> output.txt
```
This is like your first post, but there you had:
```
>> output.txt
```
.. but that _appends_ (adds) to a file if it already exists, rather than always creating a new one.

So the full command would be:
```
convert Test4x3.png txt: | sed 's/^.*#\([^ ]*\).*$/\1/g' | tr -d "\n" > output.txt
```

I suggest you find a website that explains basic bash scripting, including piping and redirection, and text filter tools like "tr" and "sed".
